### ISSUE TITLE
NGINX always connects to same coredns endpoint, even if it no longer exists

### ISSUE THREAD

**Issue by mhulscher** [cilium/cilium#27300]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

During the lifetime of an nginx process, each nginx worker (I think?) will always make UDP DNS-requests against the same CoreDNS pod IP, even if that pod no longer exists. I believe this is happening because nginx always uses the exact same source port when making DNS-requests which affects the services' backend pod that cilium selects.

The problem goes away when the nginx pod is restarted/recreated.

A workaround is to use node-local-dns with a CLRP. As long as the node-local-dns pod stays alive on the node, DNS-requests will always go there. However it becomes important to ensure that node-local-dns has started on that node *before* nginx did. Also node-local-dns can not be restarted. This is not ideal.

I was also able to reproduce this with openresty's `resolver local=on` directive.

I found out about this because I noticed that the nginx pods of my monitoring stack, which act as api-gateways (loki, grafana, tempo, mimir), stopped forwarding traffic whenever CoreDNS pods had been replaced.

![image](https://github.com/cilium/cilium/assets/18721288/93a9c130-397e-43bd-8da3-0c55468fb977)

![image](https://github.com/cilium/cilium/assets/18721288/865ae3da-e93f-4045-b5d3-a47e01733e86)

I believe this might be a similar issue in a competing CNI:

https://github.com/projectcalico/calico/issues/4509

This can be reproduced quite easily. On a cluster with cilium installed:

```
# Create namespace

kubectl create namespace nginx

# Create CiliumNetworkPolicy to allow egress

kubectl --namespace nginx apply -f- <<EOF
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: nginx
specs:
  - endpointSelector: {}
    egress:
      - toEndpoints:
          - matchLabels:
              k8s:io.kubernetes.pod.namespace: kube-system
        toPorts:
          - ports:
              - port: "53"
                protocol: ANY
            rules:
              dns:
                - matchPattern: "*"
      - toFQDNs:
          - matchName: www.example.com
        toPorts:
          - ports:
              - port: "443"
                protocol: TCP
EOF

# Install nginx

helm --namespace nginx upgrade --wait --install nginx oci://registry-1.docker.io/bitnamicharts/nginx \
  --set service.type=ClusterIP \
  --set serverBlock='
server {
  listen 0.0.0.0:8080;
  resolver kube-dns.kube-system.svc.cluster.local;
  resolver_timeout 5s;

  location / {
    set $backend www.example.com.;
    proxy_pass https://$backend:443$request_uri;
  }
}'

# Port-forward

kubectl --namespace nginx port-forward svc/nginx 8080:80 &

# Curl

( while true ; do curl -s -o /dev/null -w "%{http_code}\n" http://127.0.0.1:8080/ ; sleep 0.5 ; done ) &

# In a different window: restart coredns

kubectl -n kube-system rollout restart deploy coredns
```

After all CoreDNS pods have been replaced and the old ones no longer exist (they may still serve traffic during their lameduck period while they are Terminating):

```
200
200
200
200
200
502
502
502
502
```

### Cilium Version

1.14.0

### Kernel Version

5.10.184-175.731.amzn2.aarch64

### Kubernetes Version

v1.26.7-eks-2d98532

### Anything else?

values.yaml:

```
affinity:
  nodeAffinity:
    requiredDuringSchedulingIgnoredDuringExecution:
      nodeSelectorTerms:
      - matchExpressions:
        - key: eks.amazonaws.com/compute-type
          operator: NotIn
          values:
          - fargate
agentNotReadyTaintKey: ignore-taint.cluster-autoscaler.kubernetes.io/cilium-not-ready
bpf:
  masquerade: false
  tproxy: true
cluster:
  name: aio-lre2
cni:
  chainingMode: none
dnsProxy:
  endpointMaxIpPerHostname: 1000
  minTtl: 0
enableIPv4Masquerade: false
encryption:
  enabled: true
  nodeEncryption: true
  type: wireguard
endpointHealthChecking:
  enabled: false
endpointStatus:
  enabled: true
  status: policy
eni:
  awsEnablePrefixDelegation: true
  ec2APIEndpoint: ec2.eu-west-1.amazonaws.com
  enabled: true
  eniTags: {}
  iamRole: arn:aws:iam::630765193817:role/aio-lre2-cilium-operator
  instanceTagsFilter: aws:eks:cluster-name=aio-lre2
  updateEC2AdapterLimitViaAPI: true
envoy:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: eks.amazonaws.com/compute-type
            operator: NotIn
            values:
            - fargate
  enabled: true
  nodeSelector:
    kubernetes.io/os: linux
  priorityClassName: system-node-critical
  prometheus:
    enabled: true
    serviceMonitor:
      enabled: true
      labels:
        system: "true"
  resources:
    requests:
      cpu: 100m
      memory: 200Mi
  rollOutPods: true
  tolerations:
  - operator: Exists
  updateStrategy:
    type: OnDelete
extraConfig:
  api-rate-limit: |
    {"endpoint-create":"rate-limit:200/s,rate-burst:200,parallel-requests:200,auto-adjust:true","endpoint-delete":"rate-limit:200/s,rate-burst:200,parallel-requests:200,auto-adjust:true","endpoint-get":"rate-limit:200/s,rate-burst:200,parallel-requests:200,auto-adjust:true","endpoint-list":"rate-limit:200/s,rate-burst:200,parallel-requests:200,auto-adjust:true","endpoint-patch":"rate-limit:200/s,rate-burst:200,parallel-requests:200,auto-adjust:true"}
healthChecking: false
hubble:
  enabled: true
  eventBufferCapacity: "8191"
  metrics:
    enableOpenMetrics: true
    enabled:
    - dns:query;labelsContext=source_namespace,source_workload
    - httpV2:exemplars=true;labelsContext=source_namespace,source_workload,source_app,destination_namespace,destination_workload,destination_app,traffic_direction
    serviceMonitor:
      enabled: true
      labels:
        system: "true"
  relay:
    affinity:
      podAntiAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
        - labelSelector:
            matchExpressions:
            - key: k8s-app
              operator: In
              values:
              - hubble-relay
          topologyKey: topology.kubernetes.io/zone
    enabled: true
    podDisruptionBudget:
      enabled: true
      maxUnavailable: 1
    replicas: 3
    resources:
      requests:
        cpu: 25m
    rollOutPods: true
    updateStrategy:
      rollingUpdate:
        maxSurge: 1
        maxUnavailable: 1
      type: RollingUpdate
  ui:
    affinity:
      podAntiAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
        - labelSelector:
            matchExpressions:
            - key: k8s-app
              operator: In
              values:
              - hubble-ui
          topologyKey: topology.kubernetes.io/zone
    enabled: true
    ingress:
      className: ingress-nginx
      enabled: true
      hosts:
      - hubble.example.com
    podDisruptionBudget:
      enabled: true
      maxUnavailable: 1
    replicas: 2
    resources:
      requests:
        cpu: 25m
    rollOutPods: true
    updateStrategy:
      rollingUpdate:
        maxSurge: 1
        maxUnavailable: 1
      type: RollingUpdate
ipam:
  mode: eni
k8sServiceHost: 1EB7634C908CB343505195FC052A46C0.sk1.eu-west-1.eks.amazonaws.com
k8sServicePort: 443
kubeProxyReplacement: true
l7Proxy: true
labels: k8s:!job-name k8s:!controller-uid
loadBalancer:
  l7:
    backend: envoy
  serviceTopology: true
localRedirectPolicy: true
operator:
  affinity:
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchExpressions:
          - key: name
            operator: In
            values:
            - cilium-operator
        topologyKey: topology.kubernetes.io/zone
  extraArgs:
  - --unmanaged-pod-watcher-interval=0
  podDisruptionBudget:
    enabled: true
    maxUnavailable: 1
  priorityClassName: system-cluster-critical
  prometheus:
    enabled: true
    serviceMonitor:
      enabled: true
      labels:
        system: "true"
  replicas: 2
  resources:
    requests:
      cpu: 25m
  rollOutPods: true
  tolerations:
  - key: node.kubernetes.io/not-ready
    operator: Exists
  - key: ignore-taint.cluster-autoscaler.kubernetes.io/cilium-not-ready
    operator: Exists
  - key: ignore-taint.cluster-autoscaler.kubernetes.io/dns-not-ready
    operator: Exists
pmtuDiscovery:
  enabled: true
policyEnforcementMode: default
priorityClassName: system-node-critical
prometheus:
  enabled: true
  serviceMonitor:
    enabled: true
    labels:
      system: "true"
resources:
  requests:
    cpu: 25m
svcSourceRangeCheck: "false"
tunnel: disabled
updateStrategy:
  type: OnDelete
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by drogbeer** [comment]
This seem's quite similar to https://github.com/kubernetes/ingress-nginx/issues/9222. Whilst we initially noticed the NGINX component within our Mimir deployment requiring a restart after a cluster upgrade when CoreDNS was restarted, the more impactful for us was services that were fronted by Authentication where we specify [external authentication](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#external-authentication). 

This could be similar to the Calico in BPF mode issue you've linked, through my testing Cilium with kube-proxy is fine but when in BFP mode we see the same thing, NGINX will try use the CoreDNS pods that no longer exist when doing DNS lookups. 

With Cilium monitor we see: 

```
xx drop (Stale or unroutable IP) flow 0x29e44ff to endpoint 0, , identity 10977->unknown: 10.242.0.36:46958 -> 10.242.1.170:53 udp
xx drop (Stale or unroutable IP) flow 0x29e44ff to endpoint 0, , identity 10977->unknown: 10.242.0.36:46958 -> 10.242.1.170:53 udp
xx drop (Stale or unroutable IP) flow 0x29e44ff to endpoint 0, , identity 10977->unknown: 10.242.0.36:46958 -> 10.242.1.170:53 udp
xx drop (Stale or unroutable IP) flow 0x29e44ff to endpoint 0, , identity 10977->unknown: 10.242.0.36:46958 -> 10.242.1.170:53 udp
xx drop (Stale or unroutable IP) flow 0x29e44ff to endpoint 0, , identity 10977->unknown: 10.242.0.36:46958 -> 10.242.1.170:53 udp
xx drop (Stale or unroutable IP) flow 0x29e44ff to endpoint 0, , identity 10977->unknown: 10.242.0.36:46958 -> 10.242.1.170:53 udp
xx drop (Stale or unroutable IP) flow 0x29e44ff to endpoint 0, , identity 10977->unknown: 10.242.0.36:46958 -> 10.242.1.170:53 udp
xx drop (Stale or unroutable IP) flow 0x29e44ff to endpoint 0, , identity 10977->unknown: 10.242.0.36:46958 -> 10.242.1.170:53 udp
xx drop (Stale or unroutable IP) flow 0x29e44ff to endpoint 0, , identity 10977->unknown: 10.242.0.36:46958 -> 10.242.1.170:53 udp
xx drop (Stale or unroutable IP) flow 0x29e44ff to endpoint 0, , identity 10977->unknown: 10.242.0.36:46958 -> 10.242.1.170:53 udp
```

**Comment by mhulscher** [comment]
@drogbeer I can confirm that we were not able to reproduce this on a cluster w/o the kube-proxy replacement.

**Comment by nijich** [comment]
We observe the same problem when we redeploy our coredns with lrp in #27410. It seems that nginx use `connect()` on UDP, so the DNS requests always goes to the same endpoint. Because cilium change the dst ip with sockops DNS, the UDP socket keeps the pod' IP. When the pod gone, all the requests fail.

https://stackoverflow.com/questions/10841678/netstat-na-udp-and-state-established

**Comment by alex-berger** [comment]
Same here, and I can confirm that
- this happens with `kubeProxyReplacement=true`, `bpf-lb-sock: 'true'` and `bpf-lb-sock-hostns-only: 'false'`
- this affects all kind of Pods that use `connect(...)` with UDP sockets, not just `nginx`. Specifically, we observe this with `nginx` and JVM/Java processes, both failing (timeout) during DNS lookups.
- Disabling Socket LB by setting `bpf-lb-sock-hostns-only: 'true'` is a viable mitigation measurement and solves the problem.

Note, this has severe impact as it affects DNS lookup and causes service outages every time `coredns` is scaled or restarted. 

Maybe, we should consider adding the capability to turn off Socket LB just for UDP, while keeping it enabled for TCP.

**Comment by alex-berger** [comment]
This is related to https://github.com/cilium/cilium/issues/12859.

**Comment by chancez** [comment]
Hey all, the PR https://github.com/cilium/cilium/pull/25169 just got merged which we believe should help mitigate this particular problem.

If you have a non-prod, or dev environment available, you can try out a recent development build of Cilium with the socklb behavior. Here’s an example helm command which includes the change:

```helm install cilium -n kube-system oci://quay.io/cilium-charts-dev/cilium --version 1.15.0-dev-dev.225-main-f38e3e94a6```

Please make sure to read the updated docs for kube-proxy free (last bullet point in limitations): https://docs.cilium.io/en/latest/network/kubernetes/kubeproxy-free/#limitations

> When socket-LB feature is enabled, pods sending (connected) UDP traffic to services can continue to send traffic to a service backend even after it’s deleted. Cilium agent handles such scenarios by forcefully terminating pod sockets in the host network namespace that are connected to deleted backends, so that the pods can be load-balanced to active backends. This functionality requires these kernel configs to be enabled: CONFIG_INET_DIAG, CONFIG_INET_UDP_DIAG and CONFIG_INET_DIAG_DESTROY. If you have application pods (not deployed in the host network namespace) making long-lived connections using (connected) UDP, you can enable bpf-lb-sock-hostns-only in order to enable the socket-LB feature only in the host network namespace.

**Comment by acuteaura** [comment]
We're hitting this with Loki, Mimir and our own NGINX pods whenever AKS auto-updates coredns in the middle of the night.

**Comment by aditighag** [comment]
> Do others still have this problem?
> 
> I still have this with Cilium v1.15.4.
> 
> Edit:
> 
> Setting
> 
> ```yaml
> socketLB:
>   hostNamespaceOnly: true
> ```
> 
> in your helm values workarounds this issue.

Folks who upvoted the issue: The fix to terminate pod connections to deleted service backends with socket-LB enabled was merged to the main branch - https://github.com/cilium/cilium/pull/33459. It'll be released in the next RC (`v1.16-rc.3`), and eventually, v1.16. If you want to try the fix in the meantime, here are the images - https://github.com/cilium/cilium/actions/runs/9959326537/job/27516628767.
